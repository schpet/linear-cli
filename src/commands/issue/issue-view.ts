import { Command } from "@cliffy/command"
import { renderMarkdown } from "@littletof/charmd"
import { fetchIssueDetails, getIssueIdentifier } from "../../utils/linear.ts"
import { openIssuePage } from "../../utils/actions.ts"
import { formatRelativeTime } from "../../utils/display.ts"
import { pipeToUserPager, shouldUsePager } from "../../utils/pager.ts"
import { bold, underline } from "@std/fmt/colors"
import { getNoIssueFoundMessage } from "../../utils/vcs.ts"

export const viewCommand = new Command()
  .name("view")
  .description("View issue details (default) or open in browser/app")
  .alias("v")
  .arguments("[issueId:string]")
  .option("-w, --web", "Open in web browser")
  .option("-a, --app", "Open in Linear.app")
  .option("--no-comments", "Exclude comments from the output")
  .option("--no-pager", "Disable automatic paging for long output")
  .option("-j, --json", "Output issue data as JSON")
  .action(async (options, issueId) => {
    const { web, app, comments, pager, json } = options
    const showComments = comments !== false
    const usePager = pager !== false

    if (web || app) {
      await openIssuePage(issueId, { app, web: !app })
      return
    }

    const resolvedId = await getIssueIdentifier(issueId)
    if (!resolvedId) {
      console.error(getNoIssueFoundMessage())
      Deno.exit(1)
    }

    const issueData = await fetchIssueDetails(
      resolvedId,
      Deno.stdout.isTerminal() && !json,
      showComments,
    )

    // Handle JSON output
    if (json) {
      console.log(JSON.stringify(issueData, null, 2))
      return
    }

    const { title, description, comments: issueComments } = issueData

    let markdown = `# ${title}${description ? "\n\n" + description : ""}`

    if (Deno.stdout.isTerminal()) {
      const { columns: terminalWidth } = Deno.consoleSize()
      const renderedMarkdown = renderMarkdown(markdown, {
        lineWidth: terminalWidth,
      })

      // Capture all output in an array to count lines
      const outputLines: string[] = []

      // Add the rendered markdown lines
      outputLines.push(...renderedMarkdown.split("\n"))

      // Add comments if enabled
      if (showComments && issueComments && issueComments.length > 0) {
        outputLines.push("") // Empty line before comments
        const commentsOutput = captureCommentsForTerminal(
          issueComments,
          terminalWidth,
        )
        outputLines.push(...commentsOutput)
      }

      // Check if output exceeds terminal height and use pager if necessary
      if (shouldUsePager(outputLines, usePager)) {
        await pipeToUserPager(outputLines.join("\n"))
      } else {
        // Print directly for shorter output - same logic as pager
        outputLines.forEach((line) => console.log(line))
      }
    } else {
      if (showComments && issueComments && issueComments.length > 0) {
        markdown += "\n\n## Comments\n\n"
        markdown += formatCommentsAsMarkdown(issueComments)
      }

      console.log(markdown)
    }
  })

// Helper function to format a single comment line with consistent styling
function formatCommentHeader(
  author: string,
  date: string,
  indent = "",
): string {
  return `${indent}${underline(bold(`@${author}`))} ${
    underline(`commented ${date}`)
  }`
}

// Helper function to format comments as markdown (for non-terminal output)
function formatCommentsAsMarkdown(
  comments: Array<{
    id: string
    body: string
    createdAt: string
    user?: { name: string; displayName: string } | null
    externalUser?: { name: string; displayName: string } | null
    parent?: { id: string } | null
  }>,
): string {
  // Separate root comments from replies
  const rootComments = comments.filter((comment) => !comment.parent)
  const replies = comments.filter((comment) => comment.parent)

  // Create a map of parent ID to replies
  const repliesMap = new Map<string, typeof replies>()
  replies.forEach((reply) => {
    const parentId = reply.parent!.id
    if (!repliesMap.has(parentId)) {
      repliesMap.set(parentId, [])
    }
    repliesMap.get(parentId)!.push(reply)
  })

  // Sort root comments by creation date (newest first)
  const sortedRootComments = rootComments.slice().reverse()

  let markdown = ""

  for (const rootComment of sortedRootComments) {
    const threadReplies = repliesMap.get(rootComment.id) || []

    // Sort replies by creation date (oldest first within thread)
    threadReplies.sort((a, b) =>
      new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
    )

    const rootAuthor = rootComment.user?.displayName ||
      rootComment.user?.name ||
      rootComment.externalUser?.displayName || rootComment.externalUser?.name ||
      "Unknown"
    const rootDate = formatRelativeTime(rootComment.createdAt)

    // Format root comment
    markdown += `- **@${rootAuthor}** - *${rootDate}*\n\n`
    markdown += `  ${rootComment.body.split("\n").join("\n  ")}\n\n`

    // Format replies
    for (const reply of threadReplies) {
      const replyAuthor = reply.user?.displayName || reply.user?.name ||
        reply.externalUser?.displayName || reply.externalUser?.name ||
        "Unknown"
      const replyDate = formatRelativeTime(reply.createdAt)

      markdown += `  - **@${replyAuthor}** - *${replyDate}*\n\n`
      markdown += `    ${reply.body.split("\n").join("\n    ")}\n\n`
    }
  }

  return markdown
}

// Helper function to wrap text with proper indentation
function formatWrappedText(
  text: string,
  width: number,
  indent: string,
): string {
  const effectiveWidth = width - indent.length
  const paragraphs = text.split(/\n\s*\n/)
  const wrappedParagraphs: string[] = []

  for (const paragraph of paragraphs) {
    const words = paragraph.replace(/\s+/g, " ").trim().split(" ")
    const lines: string[] = []
    let currentLine = ""

    for (const word of words) {
      if (currentLine.length === 0) {
        currentLine = word
      } else if (currentLine.length + 1 + word.length <= effectiveWidth) {
        currentLine += " " + word
      } else {
        lines.push(indent + currentLine)
        currentLine = word
      }
    }

    if (currentLine.length > 0) {
      lines.push(indent + currentLine)
    }

    wrappedParagraphs.push(lines.join("\n"))
  }

  return wrappedParagraphs.join("\n\n")
}

// Helper function to capture comments output as string array for consistent formatting
function captureCommentsForTerminal(
  comments: Array<{
    id: string
    body: string
    createdAt: string
    user?: { name: string; displayName: string } | null
    externalUser?: { name: string; displayName: string } | null
    parent?: { id: string } | null
  }>,
  width: number,
): string[] {
  const outputLines: string[] = []

  // Separate root comments from replies
  const rootComments = comments.filter((comment) => !comment.parent)
  const replies = comments.filter((comment) => comment.parent)

  // Create a map of parent ID to replies
  const repliesMap = new Map<string, typeof replies>()
  replies.forEach((reply) => {
    const parentId = reply.parent!.id
    if (!repliesMap.has(parentId)) {
      repliesMap.set(parentId, [])
    }
    repliesMap.get(parentId)!.push(reply)
  })

  // Sort root comments by creation date (newest first)
  const sortedRootComments = rootComments.slice().reverse()

  for (const rootComment of sortedRootComments) {
    const threadReplies = repliesMap.get(rootComment.id) || []

    // Sort replies by creation date (oldest first within thread)
    threadReplies.sort((a, b) =>
      new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
    )

    const rootAuthor = rootComment.user?.displayName ||
      rootComment.user?.name ||
      rootComment.externalUser?.displayName || rootComment.externalUser?.name ||
      "Unknown"
    const rootDate = formatRelativeTime(rootComment.createdAt)

    // Format root comment using consistent styling
    outputLines.push(formatCommentHeader(rootAuthor, rootDate))
    outputLines.push(
      ...formatWrappedText(rootComment.body, width, "").split("\n"),
    )

    if (threadReplies.length > 0) {
      outputLines.push("")
    }

    // Format replies
    for (const reply of threadReplies) {
      const replyAuthor = reply.user?.displayName || reply.user?.name ||
        reply.externalUser?.displayName || reply.externalUser?.name ||
        "Unknown"
      const replyDate = formatRelativeTime(reply.createdAt)

      outputLines.push(formatCommentHeader(replyAuthor, replyDate, "  "))
      outputLines.push(
        ...formatWrappedText(reply.body, width, "  ").split("\n"),
      )
    }

    // Add spacing between comment threads, but not after the last one
    if (rootComment !== sortedRootComments[sortedRootComments.length - 1]) {
      outputLines.push("")
    }
  }

  return outputLines
}
