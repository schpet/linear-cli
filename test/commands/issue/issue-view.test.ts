import { snapshotTest } from "@cliffy/testing"
import { viewCommand } from "../../../src/commands/issue/issue-view.ts"
import { MockLinearServer } from "../../utils/mock_linear_server.ts"

// Mock the GraphQL endpoint for testing
const TEST_ENDPOINT = "http://127.0.0.1:3000/graphql"

// Common Deno args for permissions
const denoArgs = [
  "--allow-env=GITHUB_*,GH_*,LINEAR_*,NODE_ENV,EDITOR,SNAPSHOT_TEST_NAME",
  "--allow-read",
  "--allow-write",
  "--allow-run",
  "--allow-net",
  "--quiet",
]

// Test help output
await snapshotTest({
  name: "Issue View Command - Help Text",
  meta: import.meta,
  colors: false,
  args: ["--help"],
  denoArgs,
  async fn() {
    await viewCommand.parse()
  },
})

// Test with mock GraphQL endpoint
await snapshotTest({
  name: "Issue View Command - With Issue ID",
  meta: import.meta,
  colors: false,
  args: ["TEST-123"],
  denoArgs,
  async fn() {
    // Set environment variables for testing
    Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", TEST_ENDPOINT)
    Deno.env.set("LINEAR_API_KEY", "lin_api_test_key_123")

    try {
      await viewCommand.parse()
    } catch (error) {
      // Expected to fail with mock endpoint, capture the error for snapshot
      // Normalize error message to be consistent across platforms
      const message = (error as Error).message
      const normalizedMessage = message.replace(
        /Connection refused \(os error \d+\)/g,
        "Connection refused",
      )
      console.log(`Error: ${normalizedMessage}`)
    } finally {
      // Clean up environment
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})

// Test with working mock server - Terminal output (no comments available)
await snapshotTest({
  name: "Issue View Command - With Mock Server Terminal No Comments",
  meta: import.meta,
  colors: false,
  args: ["TEST-123"],
  denoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetIssueDetailsWithComments",
        variables: { id: "TEST-123" },
        response: {
          data: {
            issue: {
              title: "Fix authentication bug in login flow",
              description:
                "Users are experiencing issues logging in when their session expires. This affects the main authentication flow and needs to be resolved quickly.\n\n## Steps to reproduce\n1. Log in to the application\n2. Wait for session to expire\n3. Try to perform an authenticated action\n4. Observe the error\n\n## Expected behavior\nUser should be redirected to login page with clear messaging.\n\n## Actual behavior\nUser sees cryptic error message and gets stuck.",
              url:
                "https://linear.app/test-team/issue/TEST-123/fix-authentication-bug-in-login-flow",
              branchName: "fix/test-123-auth-bug",
              comments: {
                nodes: [],
              },
            },
          },
        },
      },
    ])

    try {
      await server.start()
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint())
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token")

      await viewCommand.parse()
    } finally {
      await server.stop()
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})

// Test with no-comments flag to disable comments
await snapshotTest({
  name: "Issue View Command - With No Comments Flag",
  meta: import.meta,
  colors: false,
  args: ["TEST-123", "--no-comments"],
  denoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetIssueDetails",
        variables: { id: "TEST-123" },
        response: {
          data: {
            issue: {
              title: "Fix authentication bug in login flow",
              description:
                "Users are experiencing issues logging in when their session expires.",
              url:
                "https://linear.app/test-team/issue/TEST-123/fix-authentication-bug-in-login-flow",
              branchName: "fix/test-123-auth-bug",
            },
          },
        },
      },
    ])

    try {
      await server.start()
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint())
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token")

      await viewCommand.parse()
    } finally {
      await server.stop()
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})

// Test with comments (default behavior)
await snapshotTest({
  name: "Issue View Command - With Comments Default",
  meta: import.meta,
  colors: false,
  args: ["TEST-123"],
  denoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetIssueDetailsWithComments",
        variables: { id: "TEST-123" },
        response: {
          data: {
            issue: {
              title: "Fix authentication bug in login flow",
              description:
                "Users are experiencing issues logging in when their session expires.",
              url:
                "https://linear.app/test-team/issue/TEST-123/fix-authentication-bug-in-login-flow",
              branchName: "fix/test-123-auth-bug",
              comments: {
                nodes: [
                  {
                    id: "comment-1",
                    body:
                      "I've reproduced this issue on staging. The session timeout seems to be too aggressive.",
                    createdAt: "2024-01-15T10:30:00Z",
                    user: {
                      name: "john.doe",
                      displayName: "John Doe",
                    },
                    externalUser: null,
                    parent: null,
                  },
                  {
                    id: "comment-2",
                    body:
                      "Working on a fix. Will increase the session timeout and add proper error handling.",
                    createdAt: "2024-01-15T14:22:00Z",
                    user: {
                      name: "jane.smith",
                      displayName: "Jane Smith",
                    },
                    externalUser: null,
                    parent: {
                      id: "comment-1",
                    },
                  },
                  {
                    id: "comment-3",
                    body:
                      "Sounds good! Also, we should add better error messaging for expired sessions.",
                    createdAt: "2024-01-15T15:10:00Z",
                    user: {
                      name: "alice.dev",
                      displayName: "Alice Developer",
                    },
                    externalUser: null,
                    parent: {
                      id: "comment-1",
                    },
                  },
                  {
                    id: "comment-4",
                    body:
                      "Should we also consider implementing automatic session refresh?",
                    createdAt: "2024-01-15T16:00:00Z",
                    user: {
                      name: "bob.senior",
                      displayName: "Bob Senior",
                    },
                    externalUser: null,
                    parent: null,
                  },
                ],
              },
            },
          },
        },
      },
    ])

    try {
      await server.start()
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint())
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token")

      await viewCommand.parse()
    } finally {
      await server.stop()
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})

// Test with mock server - Issue not found
await snapshotTest({
  name: "Issue View Command - Issue Not Found",
  meta: import.meta,
  colors: false,
  args: ["TEST-999"],
  denoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetIssueDetailsWithComments",
        variables: { id: "TEST-999" },
        response: {
          errors: [{
            message: "Issue not found: TEST-999",
            extensions: { code: "NOT_FOUND" },
          }],
        },
      },
    ])

    try {
      await server.start()
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint())
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token")

      try {
        await viewCommand.parse()
      } catch (error) {
        console.log(`Error: ${(error as Error).message}`)
      }
    } finally {
      await server.stop()
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})

// Test JSON output with no comments
await snapshotTest({
  name: "Issue View Command - JSON Output No Comments",
  meta: import.meta,
  colors: false,
  args: ["TEST-123", "--json", "--no-comments"],
  denoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetIssueDetails",
        variables: { id: "TEST-123" },
        response: {
          data: {
            issue: {
              title: "Fix authentication bug in login flow",
              description:
                "Users are experiencing issues logging in when their session expires.",
              url:
                "https://linear.app/test-team/issue/TEST-123/fix-authentication-bug-in-login-flow",
              branchName: "fix/test-123-auth-bug",
            },
          },
        },
      },
    ])

    try {
      await server.start()
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint())
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token")

      await viewCommand.parse()
    } finally {
      await server.stop()
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})

// Test JSON output with comments
await snapshotTest({
  name: "Issue View Command - JSON Output With Comments",
  meta: import.meta,
  colors: false,
  args: ["TEST-123", "--json"],
  denoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetIssueDetailsWithComments",
        variables: { id: "TEST-123" },
        response: {
          data: {
            issue: {
              title: "Fix authentication bug in login flow",
              description:
                "Users are experiencing issues logging in when their session expires.",
              url:
                "https://linear.app/test-team/issue/TEST-123/fix-authentication-bug-in-login-flow",
              branchName: "fix/test-123-auth-bug",
              comments: {
                nodes: [
                  {
                    id: "comment-1",
                    body:
                      "I've reproduced this issue on staging. The session timeout seems to be too aggressive.",
                    createdAt: "2024-01-15T10:30:00Z",
                    user: {
                      name: "john.doe",
                      displayName: "John Doe",
                    },
                    externalUser: null,
                    parent: null,
                  },
                  {
                    id: "comment-2",
                    body:
                      "Working on a fix. Will increase the session timeout and add proper error handling.",
                    createdAt: "2024-01-15T14:22:00Z",
                    user: {
                      name: "jane.smith",
                      displayName: "Jane Smith",
                    },
                    externalUser: null,
                    parent: {
                      id: "comment-1",
                    },
                  },
                ],
              },
            },
          },
        },
      },
    ])

    try {
      await server.start()
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint())
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token")

      await viewCommand.parse()
    } finally {
      await server.stop()
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})
