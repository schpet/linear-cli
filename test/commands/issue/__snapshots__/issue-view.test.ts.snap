export const snapshot = {};

snapshot[`Issue View Command - Help Text 1`] = `
stdout:
"
Usage: view [issueId]

Description:

  View issue details (default) or open in browser/app

Options:

  -h, --help     - Show this help.                           
  -w, --web      - Open in web browser                       
  -a, --app      - Open in Linear.app                        
  --no-comments  - Exclude comments from the output          
  --no-pager     - Disable automatic paging for long output  
  -j, --json     - Output issue data as JSON                 

"
stderr:
""
`;

snapshot[`Issue View Command - With Issue ID 1`] = `
stdout:
"Error: error sending request for url (http://127.0.0.1:3000/graphql): client error (Connect): tcp connect error: Connection refused
"
stderr:
"✗ Failed to fetch issue details
"
`;

snapshot[`Issue View Command - With Mock Server Terminal No Comments 1`] = `
stdout:
"# Fix authentication bug in login flow

Users are experiencing issues logging in when their session expires. This affects the main authentication flow and needs to be resolved quickly.

## Steps to reproduce
1. Log in to the application
2. Wait for session to expire
3. Try to perform an authenticated action
4. Observe the error

## Expected behavior
User should be redirected to login page with clear messaging.

## Actual behavior
User sees cryptic error message and gets stuck.
"
stderr:
""
`;

snapshot[`Issue View Command - With No Comments Flag 1`] = `
stdout:
"# Fix authentication bug in login flow

Users are experiencing issues logging in when their session expires.
"
stderr:
""
`;

snapshot[`Issue View Command - With Comments Default 1`] = `
stdout:
"# Fix authentication bug in login flow

Users are experiencing issues logging in when their session expires.

## Comments

- **@Bob Senior** - *1/15/2024*

  Should we also consider implementing automatic session refresh?

- **@John Doe** - *1/15/2024*

  I've reproduced this issue on staging. The session timeout seems to be too aggressive.

  - **@Jane Smith** - *1/15/2024*

    Working on a fix. Will increase the session timeout and add proper error handling.

  - **@Alice Developer** - *1/15/2024*

    Sounds good! Also, we should add better error messaging for expired sessions.


"
stderr:
""
`;

snapshot[`Issue View Command - Issue Not Found 1`] = `
stdout:
'Error: Issue not found: TEST-999: {"response":{"errors":[{"message":"Issue not found: TEST-999","extensions":{"code":"NOT_FOUND"}}],"status":200,"headers":{}},"request":{"query":"query GetIssueDetailsWithComments(\$id: String!) {\\\\n  issue(id: \$id) {\\\\n    title\\\\n    description\\\\n    url\\\\n    branchName\\\\n    comments(first: 50, orderBy: createdAt) {\\\\n      nodes {\\\\n        id\\\\n        body\\\\n        createdAt\\\\n        user {\\\\n          name\\\\n          displayName\\\\n        }\\\\n        externalUser {\\\\n          name\\\\n          displayName\\\\n        }\\\\n        parent {\\\\n          id\\\\n        }\\\\n      }\\\\n    }\\\\n  }\\\\n}","variables":{"id":"TEST-999"}}}
'
stderr:
"✗ Failed to fetch issue details
"
`;

snapshot[`Issue View Command - JSON Output No Comments 1`] = `
stdout:
'{
  "title": "Fix authentication bug in login flow",
  "description": "Users are experiencing issues logging in when their session expires.",
  "url": "https://linear.app/test-team/issue/TEST-123/fix-authentication-bug-in-login-flow",
  "branchName": "fix/test-123-auth-bug"
}
'
stderr:
""
`;

snapshot[`Issue View Command - JSON Output With Comments 1`] = `
stdout:
\`{
  "title": "Fix authentication bug in login flow",
  "description": "Users are experiencing issues logging in when their session expires.",
  "url": "https://linear.app/test-team/issue/TEST-123/fix-authentication-bug-in-login-flow",
  "branchName": "fix/test-123-auth-bug",
  "comments": [
    {
      "id": "comment-1",
      "body": "I've reproduced this issue on staging. The session timeout seems to be too aggressive.",
      "createdAt": "2024-01-15T10:30:00Z",
      "user": {
        "name": "john.doe",
        "displayName": "John Doe"
      },
      "externalUser": null,
      "parent": null
    },
    {
      "id": "comment-2",
      "body": "Working on a fix. Will increase the session timeout and add proper error handling.",
      "createdAt": "2024-01-15T14:22:00Z",
      "user": {
        "name": "jane.smith",
        "displayName": "Jane Smith"
      },
      "externalUser": null,
      "parent": {
        "id": "comment-1"
      }
    }
  ]
}
\`
stderr:
""
`;
